### PR TITLE
feat: dnd sortable columns in List component

### PR DESCRIPTION
**Pull request by phillipmalboeuf**
## Description

A few discussions have popped up, and clients of mine have complained, about column ordering in the list view. I've made these commits to setup dnd in the ColumnSelector component.

Features:
- Enables drag-and-drop column sorting in list view
- Standardizes dnd with new `DraggableSortable` component
- Stores sort order in user's preferences
- Deprecates `react-beautiful-dnd` for `@dnd-kit`

- [x] I have read and understand the CONTRIBUTING.md document in this repository

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Existing test suite passes locally with my changes

### CODE REVIEW

**Comment by jacobsfletch** [comment]
@phillipmalboeuf if the column selectors wrap to two lines are they still sortable via drag and drop? We have duplicative packages in Payload between `react-beautiful-dnd` and `@dnd-kit` that we need to resolved (outside of the scope of this PR). Test this out and see if it's a limitation.


Playwright looks to have drag-and-drop utilities we cna use to write tests. Check this out: https://playwright.dev/docs/input#drag-and-drop

**Comment by phillipmalboeuf** [comment]
Thanks @jacobsfletch, right at the moment they don't wrap at all, I'll see what I can do.
Looks like `react-beautiful-dnd` is being phased out, my bad, will switch to `@dnd-kit`.
And onto playwright tests as well.

**Comment by jacobsfletch** [comment]
> Thanks @jacobsfletch, right at the moment they don't wrap at all, I'll see what I can do. Looks like `react-beautiful-dnd` is being phased out, my bad, will switch to `@dnd-kit`. And onto playwright tests as well.

Not necessarily, but definitely one or the other and not both. I haven't dug too deeply into this but last I knew grid-based sorting was not possible in `react-beautiful-dnd` — if you shrink your screen they should wrap onto two lines and you could easily test this out with you current implementation.

**Comment by phillipmalboeuf** [comment]
Hey @jacobsfletch!
Wrote the e2e tests for this PR, believe everything passes.
And yes, `react-beautiful-dnd` doesn't support wrapped sorting, so if we're unhappy with the no-wrap/overflow-auto-scroll approach, we'll have to make a library change.
I'm not entirely sure if this requires a documentation change per se, what do you think?
Cheers,
Phil

**Comment by jacobsfletch** [comment]
Ok then we might want to consolidate to `@dnd-kit` for everything. And no, this feature shouldn't require any documentation change.

**Comment by phillipmalboeuf** [comment]
Hey @jacobsfletch, done!
Wrote a new `DraggableSortable` component to consolidate all the dnd elements.
The `react-beautiful-dnd` package is no longer needed, but is still installed, it felt out of bounds for me to edit the package and lock files.
Cheers!

**Comment by jacobsfletch** [comment]
@phillipmalboeuf very cool 🙌 ! I will get this reviewed.

File: src/admin/components/elements/ColumnSelector/index.tsx [hunk 1]
```diff
@@ -1,62 +1,115 @@
...
 import flattenTopLevelFields from '../../../../utilities/flattenTopLevelFields';
 import Pill from '../Pill';
 import Plus from '../../icons/Plus';
 import X from '../../icons/X';
+import DragHandle from '../../icons/Drag';
 import { Props } from './types';
 import { getTranslation } from '../../../../utilities/getTranslation';
 import { useEditDepth } from '../../utilities/EditDepth';
 import './index.scss';
+import { Field } from '../../../../fields/config/types';
```
Comment (JarrodMFlesch) [inline src/admin/components/elements/ColumnSelector/index.tsx:13]: @phillipmalboeuf do you mind bumping these above the scss import leaving a blank line between them? Just to keep consistency with other files. (I know it likely happened when you auto imported, happens all the time)

**Review by JarrodMFlesch — COMMENTED** [review]
This is awesome! We can pull it down and give it a closer look next week, hopefully Monday 😃

**Comment by JarrodMFlesch** [comment]
Looks like @jacobsfletch and I have similar feelings here 😅  Also - if  you merge with master, the access control tests that are erroring out in CI will pass.

**Comment by phillipmalboeuf** [comment]
Hey @jacobsfletch, a quick note that it's important to `setFields` as well as `setColumns` in the `moveColumn` handler.

**Comment by jacobsfletch** [comment]
> Hey @jacobsfletch, a quick note that it's important to `setFields` as well as `setColumns` in the `moveColumn` handler.

@phillipmalboeuf there's a `useEffect` in that component that calls `setFields` every time `columns` update.

**Comment by phillipmalboeuf** [comment]
> @phillipmalboeuf there's a `useEffect` in that component that calls `setFields` every time `columns` update.
That `useEffect` shouldn't be called when columns update because the `sortFields` function should only be run on `collection.fields` update.
Try dragging a field that isn't currently active and you'll see what I mean.

**Comment by jacobsfletch** [comment]
I completely agree, but we need to let that component render its given `fields` that that there is only one source of truth. To do this we need to set the column's sort index in the user's preferences.

**Comment by jacobsfletch** [comment]
I got started with this in [9445637](https://github.com/payloadcms/payload/pull/1865/commits/944563733b797b2c529c402d8e950c15582c077e)

Needs some work yet but some really good changes in there.

**Comment by jacobsfletch** [comment]
@phillipmalboeuf check it out! Column order is now stored in the user preferences. I migrated everything to a column reducer too, which has simplified this _a ton_ by moving the state logic completely away from the column selector. In this screen recording I toggle columns on and off, move some around, and refresh the browser a few times:

https://user-images.githubusercontent.com/15735305/214476321-bc1221e5-3d6e-4a8c-9c28-249d06889168.mp4

Need to get the failing tests to pass yet, but what do you think? Spin this up and give it a shot if you can!

**Comment by phillipmalboeuf** [comment]
Niiicely done @jacobsfletch! Right, to have a the columns reducer in the List component where it can trigger `setPreference` is the right move!

I do lament the loss of the "column drag handle" and "the useAsTitle link", I felt they both contributed to a better UX. I don't find the `activationConstraint` on the Pill to be reliable. And I've seen complaints in discussions that only having the leftmost column as a link to be frustrating.

(A note that I realize I'm a new contributor, and I have a lot to learn about the repo's code style and more, so thank you!)

**Comment by PatrikKozak** [comment]
Hey @phillipmalboeuf! Pulled this down and seeing some really great stuff here! There were a few things I did notice that will need to be resolved before moving ahead here.

- The sortable columns overlap each other in certain scenarios. Most likely due to the available width staying the same as the moving column. E.g. moving a smaller text column over a larger one, the larger column is shifted into the open space that still has a width of the small moving column - causing the larger column to overlap with the block to the right of that one.
- When testing the sortable columns in the drawer view, I am seeing that the columns themselves are draggable but they cannot change locations. When trying to insert, they are shifted back to their original location.
- Also noticing in the drawer view that documents are not always being populated in the list view. _No columns selected_. Maybe columns are just not enabled?

Nonetheless, great work - just a few things that will need to be looked into. I will include some video recordings below demonstrating the findings. 

https://user-images.githubusercontent.com/35232443/214639404-1c2ad389-8b2e-4832-acd5-704b98c27591.mov

https://user-images.githubusercontent.com/35232443/214639459-d3c56864-cd05-4a40-aaa4-d7776e6805a7.mov

**Comment by jacobsfletch** [comment]
@PatrikKozak THIS IS AMAZING 🙌 !

@phillipmalboeuf

> I do lament the loss of the "column drag handle" and "the useAsTitle link", I felt they both contributed to a better UX. I don't find the activationConstraint on the Pill to be reliable. And I've seen complaints in discussions that only having the leftmost column as a link to be frustrating.

I totally see your point here in both cases. One thing we need to do moving forward is achieve parity between this and relationship sorting UI as well. Related discussion here: https://github.com/payloadcms/payload/pull/1814. 

> A note that I realize I'm a new contributor,

You won't be new for very long! Thank you for taking initiative here, hopefully much more to come from you. I know we've been working together on https://github.com/payloadcms/payload/pull/1924 too.

We'll take the rest of this work internally to get it finished up—it's nearly there.

**Comment by jacobsfletch** [comment]
@PatrikKozak this is ready for another round of manual testing next chance you get!

**Review by jacobsfletch — APPROVED** [review]

**Pull request closed without merging**